site/content/blog/cue-and-carvel.md

```diff
@@ -0,0 +1,423 @@
+---
+title: "Cue and Carvel"
+slug: cue-and-carvel
+date: 2022-11-09
+author: Dmitriy Kalinin
+excerpt: "Learn how to use Cue and Carvel tools together"
+image: /img/logo.svg
+tags: ['cue', 'kapp', 'impgkg', 'kapp-controller', 'gitops', 'vendir']
+---
+
+[Cue](https://cuelang.org/) is a relatively young (but promising) programming language that enables working with data -- building data structures, validating them, querying and extracting parts. More recently you might have run into Cue being used within several tools such as Dagger (we've written about [kapp and Dagger](/blog/kapp-and-dagger/) some time ago).
+
+In this post, we'll dig into a few Cue examples for Kubernetes and see how we can use Cue and Carvel tools together. And perhaps at the end of this post, you might be interested in using Cue and Carvel as part of your Kubernetes setup...
+
+### Using Cue to build Kubernetes configuration
+
+Let's start with a simple Cue example that includes Kubernetes Deployment and Service resources:
+
+`app.cue`:
+```
+package app
+
+deployment: {
+    apiVersion: "apps/v1"
+    kind:       "Deployment"
+    metadata: {
+        namespace: "default"
+        name:      "simple-app"
+    }
+    spec: {
+        selector: matchLabels: "simple-app": ""
+        template: {
+            metadata: labels: "simple-app": ""
+            spec: containers: [{
+                name:  "simple-app"
+                image: "docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0"
+                env: [{
+                    name:  "HELLO_MSG"
+                    value: "stranger"
+                }]
+            }]
+        }
+    }
+}
+
+service: {
+    apiVersion: "v1"
+    kind:       "Service"
+    metadata: {
+        namespace: "default"
+        name:      "simple-app"
+    }
+    spec: {
+        ports: [{
+            port:       80
+            targetPort: 80
+        }]
+        selector: "simple-app": ""
+    }
+}
+```
+
+There are several things to note:
+- curly braces are explicitly used to contain maps (key-value structures)
+- brackets are explicitly used to specify collections
+- strings are always quoted which is unlike YAML where it's optional until you find ambiguity with other types like integers (this is one of my favorite choices in a configuration/data language -- i've even tried to experiment with [ytt's strict mode](https://carvel.dev/ytt/docs/v0.43.0/strict/) long ago to have a similar impact within YAML)
+- there is no notion of "documents" like in YAML -- root element of a file (and a package) is a map, hence if you want to include multiple Kubernetes resources, you'll have to find a "non-document" way to make them coexist (we'll dig into this next)
```

Comment:
> ... you'll have to find a "non-document" way ...

There's no mention of CUE's built-in support for this: `cue import` — an omission that starts to eat at the credibility of this article.

Suggestion: at a minimum, mention the feature.

https://github.com/cue-lang/cue/blob/v0.4.3/doc/tutorial/kubernetes/README.md#importing-existing-configuration